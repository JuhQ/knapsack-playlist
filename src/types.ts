export interface YoutubeItem {
  id: string;
  title: string;
  seconds: number;
  rating: number;
}
