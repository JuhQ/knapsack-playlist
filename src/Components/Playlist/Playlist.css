.playlist .ui.items > .item div,
.playlist .ui.items > .item > .content > .header:not(.ui) {
  color: white;
}
