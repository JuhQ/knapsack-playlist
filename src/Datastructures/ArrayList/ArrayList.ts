/**
 * ArrayList implementation in typescript
 */
class ArrayList<T> {
  private list: T[];

  private index: number;

  constructor(size: number) {
    this.list = Array<T>(size);
    this.index = 0;
  }

  // Time complexity O(1)
  private isFull(): boolean {
    return this.index + 1 === this.list.length;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  private grow(): T[] {
    // reserve three times as much space for the next list, so we don't need to do this operation too often
    const list = new Array<T>(this.size() * 3);

    for (let i = 0; i < this.size(); i++) {
      list[i] = this.list[i];
    }

    return list;
  }

  // Time complexity O(1), worst case O(n)
  push(value: T): void {
    if (this.isFull()) {
      this.list = this.grow();
    }

    this.list[this.index] = value;
    this.index++;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  slice(start?: number, end?: number): ArrayList<T> {
    const sliced = this.getAsArray().slice(start, end);

    const list = new ArrayList<T>(sliced.length);

    for (let i = 0; i < sliced.length; i++) {
      list.push(sliced[i]);
    }

    return list;
  }

  // remove item from the beginning of the queue and return that value
  // Time complexity O(n)
  shift(): T | undefined {
    this.index--;
    return this.list.shift();
  }

  // Time complexity O(n)
  // Space complexity O(n)
  map<K>(fn: (value: T) => K): ArrayList<K> {
    const list = this.getAsArray();
    const newList = new ArrayList<K>(this.size());

    for (let i = 0; i < list.length; i++) {
      newList.push(fn(list[i]));
    }

    return newList;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  filter(fn: (value: T) => boolean): ArrayList<T> {
    const list = this.getAsArray();
    const newList = new ArrayList<T>(1);

    for (let i = 0; i < list.length; i++) {
      if (fn(list[i])) {
        newList.push(list[i]);
      }
    }

    return newList;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  find<K>(fn: (value: T) => K): T | undefined {
    const list = this.getAsArray();

    for (let i = 0; i < list.length; i++) {
      if (fn(list[i])) {
        return list[i];
      }
    }

    return undefined;
  }

  // TODO: implement proper sort algorithm
  sort(fn: (a: T, b: T) => number): ArrayList<T> {
    const sorted = [...this.list].sort(fn);

    const list = new ArrayList<T>(this.size());
    for (let i = 0; i < this.size(); i++) {
      list.push(sorted[i]);
    }

    return list;
  }

  // Time complexity O(1)
  size(): number {
    return this.index;
  }

  // Time complexity O(1)
  at(index: number): T {
    return this.list[index];
  }

  /**
   * Return all values from the list, except undefined values
   * This will cause issue if given type would be undefined, but in this project undefined lists will never be used
   * Undefined is removed because the grow method will reserve extra space
   * Time complexity O(n)
   * Space complexity O(n)
   */
  getAsArray(): T[] {
    const list: T[] = [];
    for (let i = 0; i < this.list.length; i++) {
      if (this.list[i] !== undefined) {
        list.push(this.list[i]);
      }
    }

    return list;
  }
}

export default ArrayList;
