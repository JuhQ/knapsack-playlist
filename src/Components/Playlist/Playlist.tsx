import "./Playlist.css";

import React from "react";
import { Item } from "semantic-ui-react";

import Queue from "../../Datastructures/queue/queue";
import { YoutubeItem } from "../../types";
import PlaylistItem from "./PlaylistItem/PlaylistItem";

interface Props {
  list: Queue;
  onChange: (item: YoutubeItem) => void;
}

const Playlist: React.FC<Props> = ({ list, onChange }: Props) => (
  <div className="playlist">
    <Item.Group>
      {list.all().map((item) => (
        <PlaylistItem item={item} onClick={onChange} key={item.id} />
      ))}
    </Item.Group>
  </div>
);

export default Playlist;
