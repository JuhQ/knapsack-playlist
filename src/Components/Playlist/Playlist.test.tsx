import React from "react";
import renderer from "react-test-renderer";

import ArrayList from "../../Datastructures/ArrayList/ArrayList";
import Queue from "../../Datastructures/queue/queue";
import { YoutubeItem } from "../../types";
import Playlist from "./Playlist";

describe("Playlist", () => {
  it("should render list of songs", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60 });
    list.push({ id: "2", title: "test 2", seconds: 60 });

    const queue = new Queue();
    queue.merge(list);

    const component = renderer.create(
      <Playlist list={queue} onChange={() => null} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
