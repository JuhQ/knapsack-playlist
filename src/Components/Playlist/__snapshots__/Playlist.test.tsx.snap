// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist should render list of songs 1`] = `
<div
  className="playlist"
>
  <div
    className="ui items"
  >
    <a
      className="item"
      onClick={[Function]}
    >
      <div
        className="image"
      >
        <img
          src="https://i.ytimg.com/vi/1/hqdefault.jpg"
        />
      </div>
      <div
        className="content"
      >
        <div
          className="header"
        >
          test
        </div>
        <div
          className="meta"
        >
          1 minute
           
          
        </div>
      </div>
    </a>
    <a
      className="item"
      onClick={[Function]}
    >
      <div
        className="image"
      >
        <img
          src="https://i.ytimg.com/vi/2/hqdefault.jpg"
        />
      </div>
      <div
        className="content"
      >
        <div
          className="header"
        >
          test 2
        </div>
        <div
          className="meta"
        >
          1 minute
           
          
        </div>
      </div>
    </a>
  </div>
</div>
`;
